tests/cloudpickle_ipykernel_test.py

```diff
@@ -0,0 +1,68 @@
+import sys
+import pytest
+import textwrap
+
+from .testutils import check_deterministic_pickle
+
+if sys.platform == "win32":
+    if sys.version_info < (3, 11):
+        pytest.skip(
+            "ipykernel requires Python 3.11 or later",
+            allow_module_level=True
+        )
+ipykernel = pytest.importorskip("ipykernel")
+
+
+def run_in_notebook(code):
+    km = ipykernel.connect.jupyter_client.KernelManager()
+    km.start_kernel()
+    kc = km.client()
+    kc.start_channels()
+    status, output, err = "run", None, None
+    try:
+        assert km.is_alive() and kc.is_alive()
+        kc.wait_for_ready()
+        idx = kc.execute(code)
+        running = True
+        while running:
+            res = kc.iopub_channel.get_msg(timeout=None)
+            if res['parent_header'].get('msg_id') != idx:
+                continue
+            content = res['content']
+            if content.get("name", "state") == "stdout":
+                output = content['text']
+            if "traceback" in content:
+                err = "\n".join(content['traceback'])
+                status = "error"
+            running = res['content'].get('execution_state', None) != "idle"
+    finally:
+        kc.shutdown()
+        kc.stop_channels()
+        km.shutdown_kernel(now=True, restart=False)
+        assert not km.is_alive()
+    if status != "error":
+        status = "ok" if not running else "exec_error"
+    return status, output, err
```

Comment:
Could you please add dedicated unit tests to check that `run_in_notebook` works as expected when submitting code that raises an exception?

In particular, there are 3 kinds of possible values for `status` (`"ok"`,  `"exec_error"` and `"error"`): it seems overkill, and I am not confident that the current exception handling is safe.

Reply:
I think that's a good idea :)
Added some test for this function, not sure how to trigger a `exec_error` but I added some timeout to make sure this is correctly catched and not hanging.